File: benchmarks/BunnymarkV1Sprites/nim/BunnymarkV1Sprites.nim
import random
import godot, engine
import node_2d, texture, sprite, viewport, label, resource_loader

gdobj BunnymarkV1Sprites of Node2d:
    var texture: Texture = Texture(load("res://images/godot_bunny.png"))
    var bunnies: seq[tuple[sprite: Sprite, motion: Vector2]] = @[]
    var screenSize: Vector2
    var gravity: float = 500.0
    
    method ready*() =
        self.setProcess(true)
        
    method process*(delta: float64) =
        self.screenSize = self.getViewportRect().size

        for bunny in self.bunnies.mitems:
            var position = bunny.sprite.position
            var motion = bunny.motion

            position.x += motion.x * delta
            position.y += motion.y * delta

            motion.y += self.gravity * delta

            if position.x > self.screenSize.x:
                motion.x *= -1
                position.x = self.screenSize.x
            
            if position.x < 0:
                motion.x *= -1
                position.x = 0
            
            if position.y > self.screenSize.y:
                position.y = self.screenSize.y
                if (random(1.0) > 0.5):
                    motion.y = (random(1100.0) + 50.0)
                else:
                    motion.y *= -0.85
            
            if position.y < 0:
                motion.y = 0
                position.y = 0
            
            bunny.sprite.position = position
            bunny.motion = motion

    proc addBunny*() {.gdExport.} =
        let sprite = gdnew[Sprite]()
        sprite.texture = self.texture
        self.addChild(sprite)
        sprite.position = vec2(self.screenSize.x / 2, self.screenSize.y / 2)
        self.bunnies.add((sprite, vec2(random(200.0) + 50.0, random(200.0) + 50.0)))
    
    proc removeBunny*() {.gdExport.} =
        if self.bunnies.len == 0:
            return

        var bunny = self.bunnies.pop();
        self.removeChild(bunny.sprite)
    
    proc finish*() {.gdExport.} =
        self.emitSignal("benchmark_finished", @[newVariant(self.bunnies.len)])

File: benchmarks/BunnymarkV1Sprites/es/BunnymarkV1Sprites.js
class BunnymarkV1Sprites extends godot.Node2D {
	bunnies = [];
	grav = 500;
	bunny_texture = godot.ResourceLoader.load("res://images/godot_bunny.png");
	screen_size = null;
	bunnies_container = new godot.Node2D();
	
	_ready() {
		this.add_child(this.bunnies_container);
	}

	_process(delta) {
		this.screen_size = this.get_viewport_rect().size;
		for (let i = 0 ; i < this.bunnies.length; i++) {
			let bunny = this.bunnies[i];
			let pos = bunny.sprite.position;
			let newPosition = bunny.pos;
			
			pos.x += newPosition.x * delta;
			pos.y += newPosition.y * delta;
			newPosition.y += this.grav * delta;
		
			if (pos.x > this.screen_size.x) {
				newPosition.x *= -1;
				pos.x = this.screen_size.x;
			}
			
			if (pos.x < 0) {
				newPosition.x *= -1;
				pos.x = 0;
			}
				
			if (pos.y > this.screen_size.y) {
				pos.y = this.screen_size.y;
				if (Math.random() > 0.5)
					newPosition.y = -(godot.randi() % 1100 + 50);
				else
					newPosition.y *= -0.85;
			}
				
			if (pos.y < 0) {
				newPosition.y = 0;
				pos.y = 0;
			}
			
			bunny.sprite.position = pos;
			bunny.pos = newPosition;
		}
	}
	
	add_bunny() {
		let bunny = new godot.Sprite();
		bunny.set_texture(this.bunny_texture);
		this.bunnies_container.add_child(bunny);
		bunny.position = new godot.Vector2(this.screen_size.x / 2, this.screen_size.y / 2);
		this.bunnies.push({
			sprite: bunny,
			pos: new godot.Vector2(godot.randi() % 200 + 50, godot.randi() % 200 + 50)
		});
	}
	
	remove_bunny() {
		if (this.bunnies.length == 0)
			return;
		var bunny = this.bunnies.pop();
		bunny.sprite.queue_free();
	}
	
	get_bunny_count() {
		return this.bunnies.length;
	}
	
	finish() {
		this.emit_signal("benchmark_finished", this.bunnies.length);
	}

}

godot.register_class(BunnymarkV1Sprites);
